### PR TITLE
temperature-sensor

### PR DESCRIPTION
**Pull request by godstack**

### CODE REVIEW

File: client/package.json [hunk 1]
```diff
@@ -13,13 +13,12 @@
     "not ie < 9"
   ],
   "dependencies": {
+    "axios": "^0.19.0",
```
Comment (undead404) [inline client/package.json]: Please remove the caret as it's an unstable dependency

File: client/package.json [hunk 2]
```diff
@@ -13,13 +13,12 @@
...
     "body-parser": "~1.19.0",
     "bootstrap": "~4.3.1",
-    "chart.js": "~2.8.0",
-    "classnames": "~2.2.6",
-    "cookie-parser": "~1.4.4",
+    "chart.js": "^2.8.0",
```
Comment (undead404) [inline client/package.json]: Replace the caret with the tilde (`~`) in stable dependencies

File: client/src/routes/temperaturesensor/chart-dataset.js [hunk 3]
```diff
@@ -0,0 +1,34 @@
+import format from 'moment';
```
Comment (undead404) [inline client/src/routes/temperaturesensor/chart-dataset.js]: this library is called `moment`, not `format`, rename your import please

File: client/src/routes/temperaturesensor/temperature.js [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+              checked={this.parseStatus(info.Status)}
+              onClick={this.statusOnClick(info.Status)}
+            >
+              {this.text}
+            </button>
+            {/* <Icon text={this.text} /> */}
+            <h1>{events.degree}</h1>
+          </div>
+        </div>
+        {/* <Icon  /> */}
```
Comment (undead404) [inline client/src/routes/temperaturesensor/temperature.js]: Please remove this garbage comments

File: client/src/routes/temperaturesensor/temperature.js [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+
+  statusOnClick = status => {
+    return () => this.handleOnClick(status);
+  };
+
+  parseStatus = status => {
+    switch (status) {
+      case 0:
+        this.text = 'ON';
+        startButton = 'btn btn-success';
```
Comment (undead404) [inline client/src/routes/temperaturesensor/temperature.js]: Funny trick. You have to derive your attributes from props without unnecessary intermediate variables

File: generator/temperature/main.go [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+func createTemperatureEvent() {
+	s1 := rand.NewSource(time.Now().UnixNano())
+	r1 := rand.New(s1)
+	min := 18
+	max := 28
+
+	degree := r1.Intn(max-min+1) + min
+	step := 1
+	for {
+		moveStep := r1.Intn(3)
```
Comment (undead404) [inline generator/temperature/main.go]: Please use `NormFloat64` here

File: server/domain/devices/temperature.go [hunk 7]
```diff
@@ -24,17 +24,42 @@ func init() {
...
 	}
-
 	return device, err
 }
 
+func (t *Temperature) EventFilter(from string) ([]TemperatureEvent, error) {
+	var events []TemperatureEvent
+
+	var err error
+
```
Comment (undead404) [inline server/domain/devices/temperature.go:65]: Please cast `from` into a time object and proceed only if no errors arise

**Review by undead404 — CHANGES_REQUESTED** [review]

File: .gitignore [hunk 8]
```diff
@@ -30,5 +30,6 @@ yarn-error.log*
 ./server/go.sum
 **/*.exe
 server/server
-generator/
-!generator/**/*.go
+server/server.exe
+generator/temperature/temperature.exe
```
Comment (undead404) [inline .gitignore:37]: Please restore that removed lines

File: client/package.json [hunk 9]
```diff
@@ -13,14 +13,14 @@
     "not ie < 9"
   ],
   "dependencies": {
-    "axios": "0.19.0",
+    "axios": "~0.19.0",
```
Comment (undead404) [inline client/package.json]: No tilde in unstable dependencies please, restore it as in master

File: client/package.json [hunk 10]
```diff
@@ -13,14 +13,14 @@
...
     "body-parser": "~1.19.0",
     "bootstrap": "~4.3.1",
     "chart.js": "~2.8.0",
     "classnames": "~2.2.6",
-    "cookie-parser": "~1.4.4",
+    "cookie-parser": "^1.4.4",
```
Comment (undead404) [inline client/package.json]: No caret, only tilde! (for stable only)

File: client/src/routes/temperaturesensor/temperature.js [hunk 11]
```diff
@@ -0,0 +1,217 @@
...
+      dispatchChangeStatus(true);
+    }
+  };
+
+  render() {
+    const { events, info, isLoading, error } = this.props;
+
+    // const lastEvent = events.pop().Degree;
+
+    if (error) {
```
Comment (undead404) [inline client/src/routes/temperaturesensor/temperature.js]: So if you have error, everything else disappears? That's wrong, please show your error above or below your chart

File: generator/temperature/main.go [hunk 12]
```diff
@@ -0,0 +1,122 @@
+package main
```
Comment (undead404) [inline generator/temperature/main.go]: Please change your code to match what Artem did to run all generators simultaneously

File: go.sum [hunk 13]
```diff
@@ -1,7 +1,10 @@
 cloud.google.com/go v0.26.0/go.mod h1:aQUYkXzVsufM+DwF1aE+0xfcU+56JwCaLick0ClmMTw=
```
Comment (undead404) [inline go.sum]: Please delete this file, run all `go` things in `server` dir

**Review by undead404 — CHANGES_REQUESTED** [review]

File: go.mod [hunk 14]
```diff
@@ -1,132 +0,0 @@
-module github.com/elifTech/school-project-2019
```
Comment (undead404) [inline go.mod]: Please delete `go.sum` instead

**Review by undead404 — CHANGES_REQUESTED** [review]

File: yarn.lock [hunk 15]
```diff
@@ -2,7 +2,3 @@
 # yarn lockfile v1
```
Comment (undead404) [inline yarn.lock]: Remove this file pls

**Review by undead404 — CHANGES_REQUESTED** [review]

File: server/domain/routers/temperature.go [hunk 16]
```diff
@@ -4,30 +4,58 @@ import (
...
-	router.GET("/temperature/ping", PingTemperature)
+func enableCors(w *http.ResponseWriter) {
+	(*w).Header().Set("Access-Control-Allow-Origin", "*")
+}
+
+func FilterTemperatureEvents(w http.ResponseWriter, r *http.Request, _ httprouter.Params) {
+	keys := r.URL.Query()
+	from := keys.Get("from")
+	temperature := devices.Temperature{}
+	temperatureEvent, err := temperature.EventFilter(from)
```
Comment (undead404) [inline server/domain/routers/temperature.go]: parse time please before proceeding

**Review by undead404 — CHANGES_REQUESTED** [review]

File: client/src/routes/index.js [hunk 17]
```diff
@@ -31,10 +31,17 @@ const routes = {
...
+        import(
+          /* webpackChunkName: 'temperaturesensor' */ './temperaturesensor'
+        ),
+      path: '/temperaturesensor',
+    },
     {
       load: () =>
         import(/* webpackChunkName: 'carbonsensor' */ './carbonsensor'),
-      path: '/carbonmonoxide',
+      path: '/carbonsensor',
```
Comment (undead404) [inline client/src/routes/index.js]: Does it work in your branch after merging?

File: client/src/routes/temperaturesensor/temperature.js [hunk 18]
```diff
@@ -0,0 +1,299 @@
...
+  setFilter = (filter, period) => {
+    return () => this.getFilterData(filter, period);
+  };
+
+  setDateFormat = date => {
+    return moment(date).format('YYYY-MM-DD HH:mm:ss');
+  };
+
+  getFilterData = (filter, period) => {
+    this.selectedButton = filter;
```
Comment (undead404) [inline client/src/routes/temperaturesensor/temperature.js]: Please use local state instead

File: client/src/routes/temperaturesensor/temperature.js [hunk 19]
```diff
@@ -0,0 +1,299 @@
...
+
+  statusOnClick(status) {
+    return () => this.handleOnClick(status);
+  }
+
+  parseStatus = status => {
+    switch (status) {
+      case 0:
+        this.text = 'ON';
+        this.startButton = 'btn btn-success';
```
Comment (undead404) [inline client/src/routes/temperaturesensor/temperature.js]: Please use local state instead or derive these values directly from props

File: client/src/routes/temperaturesensor/temperatureIcon/Icons/cold.js [hunk 20]
```diff
@@ -0,0 +1,83 @@
+import React from 'react';
+
+const iconCold = (
+  <svg
```
Comment (undead404) [inline client/src/routes/temperaturesensor/temperatureIcon/Icons/cold.js]: Isn't it static? Why don't you store it as plain SVG?

File: client/src/routes/temperaturesensor/temperatureIcon/Icons/hot.js [hunk 21]
```diff
@@ -0,0 +1,83 @@
+import React from 'react';
+
+const iconHot = (
```
Comment (undead404) [inline client/src/routes/temperaturesensor/temperatureIcon/Icons/hot.js]: Isn't it static? Why don't you store it as plain SVG?

File: client/src/routes/temperaturesensor/temperatureIcon/Icons/normal.js [hunk 22]
```diff
@@ -0,0 +1,155 @@
+import React from 'react';
+
+const iconNormal = (
+  <svg
```
Comment (undead404) [inline client/src/routes/temperaturesensor/temperatureIcon/Icons/normal.js]: Isn't it static? Why don't you store it as plain SVG?

**Review by undead404 — CHANGES_REQUESTED** [review]

File: server/domain/devices/temperature.go [hunk 23]
```diff
@@ -39,24 +37,43 @@ func (t *Temperature) Get() (*Temperature, error) {
...
+	return device, err
+}
+
+func (t *Temperature) EventFilter(from string) ([]TemperatureEvent, error) {
+	var events []TemperatureEvent
+
+	var err error
+
+	err = Storage.Where("created_at > ?", from).Order("created_at").Find(&events).Error
+	fmt.Println(events)
```
Comment (polluxx) [inline server/domain/devices/temperature.go:67]: Remove all prints please

File: eventGenerator/temperature.go [hunk 24]
```diff
@@ -0,0 +1,115 @@
...
+func toFixed(num float64, precision int) float64 {
+	output := math.Pow(10, float64(precision))
+	return float64(round(num*output)) / output
+}
+
+func round(num float64) int {
+	return int(num + math.Copysign(0.5, num))
+}
+
+func GenerateTemperatureEvent() {
```
Comment (polluxx) [inline eventGenerator/temperature.go]: Please use generator, which we have in main;
Do NOT use an infinite loop for such things

File: server/domain/routers/temperature.go [hunk 25]
```diff
@@ -7,24 +7,51 @@ import (
...
+		return
+	}
 
-	router.GET("/temperature/all", AllTemperature)
+	response, err := json.Marshal(temperatureEvent)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusNotFound)
+		return
+	}
+	w.Header().Set("Content-Type", "application/json")
```
Comment (polluxx) [inline server/domain/routers/temperature.go:57]: you can remove that if you are using cors stuff below

File: server/domain/routers/temperature.go [hunk 26]
```diff
@@ -38,16 +65,67 @@ func PingTemperature(w http.ResponseWriter, r *http.Request, ps httprouter.Param
 		http.Error(w, err.Error(), http.StatusNotFound)
 		return
 	}
+	enableCors(&w)
```
Comment (polluxx) [inline server/domain/routers/temperature.go:75]: Is that even builds?

File: client/src/routes/temperaturesensor/temperature.js [hunk 27]
```diff
@@ -0,0 +1,303 @@
...
+defaults.global.defaultFontFamily = 'Montserrat';
+
+const buttonsStyle = {
+  // height: '150px',
+  marginLeft: '9em',
+  marginTop: '1em',
+  // width: '200px',
+};
+
+const defaultButt = 'btn btn-secondary btn-sm';
```
Comment (undead404) [inline client/src/routes/temperaturesensor/temperature.js]: rename please https://www.dictionary.com/browse/butt

File: client/src/routes/temperaturesensor/temperature.js [hunk 28]
```diff
@@ -0,0 +1,303 @@
...
+      dispatchChangeStatus(true);
+    }
+  };
+
+  render() {
+    const { events, info, isLoading, error } = this.props;
+    let { isVisible } = this.props;
+
+    if (events.length !== 0) {
+      this.degree = events.slice(-1)[0].degree;
```
Comment (undead404) [inline client/src/routes/temperaturesensor/temperature.js]: Why do you bind this to `this`? Make it a local constant

File: client/src/routes/temperaturesensor/temperature.js [hunk 29]
```diff
@@ -0,0 +1,303 @@
...
+      const degreeArray = [];
+      /* eslint-disable no-restricted-syntax */
```
Comment (undead404) [inline client/src/routes/temperaturesensor/temperature.js]: Remove this directive and use `forEach` or `map` instead.

File: eventGenerator/temperature.go (hunk 24 above)
Comment (undead404) [reply eventGenerator/temperature.go]: I'd add if you need to derive next value from a previous one - store that previous value in the upper scope

**Review by undead404 — CHANGES_REQUESTED** [review]

File: eventGenerator/temperature.go [hunk 30]
```diff
@@ -0,0 +1,248 @@
...
+func toFixed(num float64, precision int) float64 {
+	output := math.Pow(10, float64(precision))
+	return float64(round(num*output)) / output
+}
+
+func round(num float64) int {
+	return int(num + math.Copysign(0.5, num))
+}
+
+// func checkSensorsStatus() error {
```
Comment (undead404) [inline eventGenerator/temperature.go]: Please remove this garbage

**Review by undead404 — CHANGES_REQUESTED** [review]

File: client/src/routes/temperaturesensor/temperature.js [hunk 31]
```diff
@@ -0,0 +1,313 @@
...
+  marginLeft: '5.5em',
+  marginTop: '1em',
+};
+
+const xAxeStyle = {
+  font: '20px serif',
+  paddingLeft: '17em',
+};
+
+const defaultButt = 'btn btn-secondary btn-lg';
```
Comment (undead404) [inline client/src/routes/temperaturesensor/temperature.js]: Rename this. https://context.reverso.net/%D0%BF%D0%B5%D1%80%D0%B5%D0%B2%D0%BE%D0%B4/%D0%B0%D0%BD%D0%B3%D0%BB%D0%B8%D0%B9%D1%81%D0%BA%D0%B8%D0%B9-%D1%80%D1%83%D1%81%D1%81%D0%BA%D0%B8%D0%B9/butt

**Review by undead404 — CHANGES_REQUESTED** [review]

File: client/src/routes/temperaturesensor/temperature.js [hunk 32]
```diff
@@ -0,0 +1,314 @@
...
+                }
+                name="hours"
+                onClick={this.setFilter('hours', 2)}
+              >
+                Last Hours
+              </button>
+              <button
+                type="button"
+                className={
+                  this.selectedButton === 'days' ? activeButton : defaultButton
```
Comment (undead404) [inline client/src/routes/temperaturesensor/temperature.js]: please derive your classes with `classnames` instead

Comment (undead404) [reply client/src/routes/temperaturesensor/temperature.js]: And in other places also

File: client/src/routes/temperaturesensor/temperature.js (hunk 18 above)
Comment (undead404) [reply client/src/routes/temperaturesensor/temperature.js]: This line should be removed.

File: client/src/routes/temperaturesensor/temperatureIcon/temperature-icon.js.rej [hunk 33]
```diff
@@ -0,0 +1,3 @@
+diff a/client/src/routes/temperaturesensor/temperatureIcon/temperature-icon.js b/client/src/routes/temperaturesensor/temperatureIcon/temperature-icon.js	(rejected hunks)
```
Comment (undead404) [inline client/src/routes/temperaturesensor/temperatureIcon/temperature-icon.js.rej]: It's definitely a trashy file

**Review by undead404 — CHANGES_REQUESTED** [review]

File: client/package.json [hunk 34]
```diff
@@ -17,7 +17,7 @@
     "body-parser": "~1.19.0",
     "bootstrap": "~4.3.1",
     "chart.js": "~2.8.0",
-    "chartjs-plugin-labels": "~1.1.0",
+    "chartjs-plugin-labels": "^1.1.0",
```
Comment (undead404) [inline client/package.json]: Please restore tilde here

File: client/src/routes/temperaturesensor/temperature.js [hunk 35]
```diff
@@ -0,0 +1,314 @@
...
+  marginLeft: '5.5em',
+  marginTop: '1em',
+};
+
+const xAxeStyle = {
+  font: '20px serif',
+  paddingLeft: '17em',
+};
+
+const defaultButton = 'btn btn-secondary btn-lg';
```
Comment (undead404) [inline client/src/routes/temperaturesensor/temperature.js]: Please remove this and next line

File: client/src/routes/temperaturesensor/temperature.js [hunk 36]
```diff
@@ -0,0 +1,314 @@
...
+      Name: PropTypes.string.isRequired,
+      Status: PropTypes.number.isRequired,
+      Type: PropTypes.string.isRequired,
+    }).isRequired,
+    isLoading: PropTypes.bool.isRequired,
+    isVisible: PropTypes.bool.isRequired,
+    removeInterval: PropTypes.func.isRequired,
+  };
+
+  componentDidMount() {
```
Comment (undead404) [inline client/src/routes/temperaturesensor/temperature.js]: remove this

**Review by undead404 — CHANGES_REQUESTED** [review]

File: server/domain/routers/temperature.go (hunk 26 above)
Comment (godstack) [reply server/domain/routers/temperature.go:75]: yes, it works

**Review by undead404 — APPROVED** [review]

**Review by undead404 — APPROVED** [review]

File: eventGenerator/package.json [hunk 37]
```diff
@@ -0,0 +1,5 @@
+{
+  "dependencies": {
+    "isomorphic-cookie": "^1.2.4"
```
Comment (undead404) [inline eventGenerator/package.json]: Remove this garbage file please

File: eventGenerator/yarn.lock [hunk 38]
```diff
@@ -0,0 +1,21 @@
+# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
+# yarn lockfile v1
+
+
+cookie@^0.1.2:
```
Comment (undead404) [inline eventGenerator/yarn.lock]: Remove this garbage file please

File: package.json [hunk 39]
```diff
@@ -0,0 +1,5 @@
+{
+  "dependencies": {
+    "isomorphic-cookie": "^1.2.4"
```
Comment (undead404) [inline package.json]: Remove this garbage file please

**Review by undead404 — CHANGES_REQUESTED** [review]

File: client/src/routes/temperaturesensor/TemperatureWidget/TemperatureWidget.js [hunk 40]
```diff
@@ -0,0 +1,130 @@
...
+    let degree;
+    let createdAt;
+    if (events.length !== 0) {
+      degree = events.slice(-1)[0].degree;
+      createdAt = moment(events.slice(-1)[0].CreatedAt).format(
+        'DD-MM HH:mm:ss',
+      );
+    }
+
+    dispatchWidgetData({
```
Comment (undead404) [inline client/src/routes/temperaturesensor/TemperatureWidget/TemperatureWidget.js]: why do you dispatch it in render? Do you understand that dispatching an action can cause an additional render? Consider moving this call to constructor or `componentDidMount` (or `componentDidUpdate` if you are crazy)

File: client/src/routes/temperaturesensor/TemperatureWidget/TemperatureWidget.js [hunk 41]
```diff
@@ -0,0 +1,130 @@
...
+        </div>
+      </div>
+    );
+  }
+
+  statusOnClick(status) {
+    return () => this.handleOnClick(status);
+  }
+
+  checkStatus = status => {
```
Comment (undead404) [inline client/src/routes/temperaturesensor/TemperatureWidget/TemperatureWidget.js]: Yeah, I told about `parseStatus` and you just renamed it? Once again, why this is a method not utilizing `this`?

**Review by undead404 — CHANGES_REQUESTED** [review]

File: client/src/routes/temperaturesensor/TemperatureWidget/TemperatureWidget.js [hunk 42]
```diff
@@ -0,0 +1,129 @@
...
+        )}
+        <div className={style.innerContainer}>
+          <Row>
+            <Col md={8} className={style.header}>
+              Temperature Sensor
+            </Col>
+            <Col md={2} className="ml-4">
+              <Switch
+                onChange={this.statusOnClick(info.Status)}
+                checked={info.Status === 1 ? false : info.Status === 0}
```
Comment (undead404) [inline client/src/routes/temperaturesensor/TemperatureWidget/TemperatureWidget.js]: What's this? Isn't it simpler to write just `info.Status !==1` instead?

**Review by undead404 — CHANGES_REQUESTED** [review]

File: eventGenerator/main.go [hunk 43]
```diff
@@ -24,8 +24,8 @@ func generate(seconds int, functions ...func()) {
 
 func main() {
 	gotenv.Load()
-	generate(5, GenerateWaterQualityEvent, GenerateWindEvent, GenerateCarbonEvent)
-	generate(60, GenerateWaterConsumptionEvent)
+	generate(5, GenerateWaterQualityEvent, GenerateWindEvent, GenerateCarbonEvent, GenerateTemperatureEvent)
+	generate(60, GenerateWaterMeterEvent)
```
Comment (undead404) [inline eventGenerator/main.go]: Does this build? `GenerateWaterMeterEvent` doesn't exist anymore, I suppose, it's renamed

**Review by undead404 — CHANGES_REQUESTED** [review]

**Review by undead404 — APPROVED** [review]

**Review by undead404 — APPROVED** [review]

**Review by undead404 — APPROVED** [review]

**Pull request merged**